libraries/chain/database.cpp

```diff
@@ -1132,7 +1132,7 @@ asset database::create_vesting( const account_object& to_account, asset liquid,
          FC_ASSERT( liquid.symbol.is_vesting() == false );
          // Get share price.
          const auto& smt = get< smt_token_object, by_symbol >( liquid.symbol );
-         FC_ASSERT( smt.allow_voting == to_reward_balance, "No voting - no rewards" );
+         FC_ASSERT( smt.allow_voting || to_reward_balance == false, "No voting - no rewards" );
```

Comment:
I think this assert is still problematic, but for a different reason than stated in #2253

The vesting payout code isn't done yet, but it appears that this is designed for the payout code to blindly call `create_vesting` and fail if there shouldn't be a payout. This is bad way to implement this when the state of allowed voting can be determined before `create_vesting` is called. As a general rule, we only throw exceptions in consensus code to indicate a consensus rule is broken and the operation/transaction/block in question should be rejected. In this case it appears that you are using the exception to indicate a control path, which is not a good use of exceptions.

Reply:
OK, I replaced the assertion with zero amount return. See commit https://github.com/steemit/steem/pull/2356/commits/ca85d02645d2477217f58702e1ade07b66e24bdc
My reasoning is that though putting that assertion in `create_vesting` was a bad idea for reasons you described, but still it is a good place to check SMT's `allow_voting` property and react appropriately, because a) we've got the smt token object in hand and b) it allows to keep the STEEM vs SMT ifology hidden inside the method.